TFUtil.py

```diff
@@ -8517,3 +8517,177 @@ def get_non_deterministic_ops_from_graph():
     # elif ... more non det ops to be added
 
   return non_det_ops
+
+
+def sum_rows(x):
+  """
+  Returns a vector summing up each row of the matrix x.
+
+  This is mostly a copy of
+    https://github.com/tensorflow/tensorflow/blob/e19c354920c3b246dda6598229210a582caaa1a9/tensorflow/python/ops/nn_impl.py#L1427
+
+  :param tf.Tensor x: input tensor
+  :return: Vector, which rows are summed along axis 1 with shape [shape(x)[0]]
+  :rtype: tf.Tensor
+  """
+  # sum_rows(x) is equivalent to math_ops.reduce_sum(x, 1) when x is
+  # a matrix.  The gradient of _sum_rows(x) is more efficient than
+  # reduce_sum(x, 1)'s gradient in today's implementation. Therefore,
+  # we use _sum_rows(x) in the nce_loss() computation since the loss
+  # is mostly used for training.
```

Comment:
`sum_rows` or `_sum_rows`? Be consistent.
Also, maybe this description would actually be also helpful in the docstring itself, not just as a code comment? So move it there.

Reply:
Btw, are you sure that the gradient is more efficient? Have you measured that? Or do they write that in the TF comment?
I remember seeing people using `matmul` instead of `reduce_sum`, but the reason was that in earlier TF versions, `matmul` was deterministic and `reduce_sum` was not (in later TF versions, also `reduce_sum` should be deterministic).

Reply:
I just checked; that comment is copied from the TF code. So they claim that. But that code is at least 4 years old, probably even older, and dates back to before TF was public (see [here](https://github.com/tensorflow/tensorflow/blame/ef2a926ec05dfd337d84279aafa58b22f0f36123/tensorflow/python/ops/nn.py#L1084)). So I'm not sure if that is still true today. I would verify that if you have time (just try out with `tf.reduce_sum` instead, and see if there is a difference). In any case, replace "today", and be more specific (write the date of the initial TF commit).

Reply:
I see, thanks.
I removed sum_rows now and replaced this with reduce_sum. I did some small scale tests and on the V100 there is no difference in runtime.